src/client/executor.rs

```diff
@@ -397,6 +382,29 @@ impl Client {
                     implicit_session,
                 },
                 Err(mut err) => {
+                    // If the error is a reauthentication required error, we reauthenticate and retry the operation.
+                    if err.is_reauthentication_required() {
+                        let credential = self.inner.options.credential.as_ref().ok_or(
+                            ErrorKind::Authentication {
+                                message:
+                                    "No Credential when reauthentication required error occured"
+                                        .to_string(),
+                            },
+                        )?;
+                        let server_api = self.inner.options.server_api.as_ref();
+
+                        credential
+                            .mechanism
```

Comment:
AuthMechanism should really be `Copy`, but it's public, so we wouldn't want to change that as it's an API breaking change.